### ISSUE TITLE
Allow custom partitioning for Generic Text Editor 

### ISSUE THREAD

**Issue by zulus** [eclipse-platform/eclipse.platform.ui#1359]
## Suggestion

Currently Generic Text Editor always works with default partitioner (ExtensionBasedTextViewerConfiguration.getConfiguredDocumentPartitioning isn't touched).  Everything is always connected to default partitioner (and all legal types).  Replacing "__dftl_partitioning" off course trigger warning in console.

Would be nice to have ability to register own partitioning, and later in extensions bind *Stategy/*Providers/*Processors to concrete partition.

For example I need partitioner to separate PHP (embeded language) from rest(html), and introduce different extensions for each partition (probably somehow forward to LSP4E non-php regions). Without above, I'll have to introduce composites and re-run partitioning for each extension

## Community

- [x] I understand suggesting an enhancement doesn't mandate anyone to implement it. Other contributors may consider this suggestion, or not, at their own convenience. The most efficient way to get it fixed is that I implement it myself and contribute it back as a good quality patch to the project.

**Comment by laeubi** [comment]
@zulus I'm currently working on a similar thing and its quite hard to find out things as there is no general document about all possible options, but you can use the [extension described here](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/guide/editors_documents.htm?cp=2_0_13_2) to install a [IDocumentSetupParticipant](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/api/org/eclipse/core/filebuffers/IDocumentSetupParticipant.html) where then you can install additional partition by checking the document for the interface `IDocumentExtension3`.

What I currently have not found out is how to register an extension for a partition, it seems these are always registered for a document type and then must internally handle if the current partition is of interest for them, e.g but using the current offset in an autocomplete.

**Comment by mickaelistria** [comment]
Indeed, @laeubi is right: partitionning isn't really a duty of the editor, but of the document; and using documentSetupParticipant should be the right place where to hook partitioning to an existing document, whichever editor uses it.
Then adding as **optional** field to extensions the target partition would be welcome.

**Comment by zulus** [comment]
@mickaelistria yes and now because TextSourceViewerConfiguration return id of partitioner that should by used by Editor (it's source viewer). If I replace default partitioner via IDocumentSetupParticipant , warning in logs appear. If I setup own, generic editor extensions ignore it

For now generic editor and it's composite reconcilers/processors are hardcoded to IDocumentExtension3.DEFAULT_PARTITIONING. For example GenericEditorContentAssistant always register processors to all IDocument.getLegalContentTypes (version that use default_partitioning only)

**Comment by laeubi** [comment]
> If I replace default partitioner via IDocumentSetupParticipant , warning in logs appear

No one said you should replace the default paritioner!

>  For example GenericEditorContentAssistant always register processors to all IDocument.getLegalContentTypes (version that use default_partitioning only)

That's what I said, you then need to check inside your content assitant what partion you are actually in (what works as I have tested already), and what @mickaelistria means that it would be good to enhance generic editor to maybe restrict it to a given partion type on the first place, I could think about simply add a new attribute for the ext point to specify the paritioning type.
